File: internal/expr/binary.go
package expr

import "github.com/aland20/interpreter/internal"

type Binary struct {
	Left     Expr
	Operator internal.Token
	Right    Expr
}

func NewBinary(left Expr, operator internal.Token, right Expr) Expr {
	return &Binary{
		Left:     left,
		Operator: operator,
		Right:    right,
	}
}

func (b *Binary) Accept(v Visitor) any {
	return v.VisitBinaryExpr(b)
}

File: src/expr/binary.rs
use crate::token::Token;

use super::{Expr, ExprVisitor, Visitor};

pub struct Binary {
    pub left: Box<dyn Expr>,
    pub operator: Token,
    pub right: Box<dyn Expr>,
}

impl Binary {
    pub fn new(left: Box<dyn Expr>, operator: Token, right: Box<dyn Expr>) -> Self {
        Binary {
            left,
            operator,
            right,
        }
    }
}

impl Visitor<&dyn ExprVisitor<()>> for Binary {
    fn accept(&self, visitor: &dyn ExprVisitor<()>) {
        visitor.visit_binary_expr(self);
    }
}
